### DISCUSSION TITLE
Input/output pc to Technic hub

### DISCUSSION THREAD

**Discussion by Devilly** [pybricks/support#1148] [Q&A] [answered]
Hi,

I made some code where the Inventor hub is started and waits for input from the pc. Once this is given the hub uses the given input for the rest of the execution.

This works perfectly fine but when using the same code, only then with the different hub constructor, with a Technic hub it doesn't seem to receive the data from the pc.

Is there any difference I should be aware of between this functionality on both hubs?

Regards

**Comment by laurensvalk** [comment]
Can you share your code? I think this functionality should work the same across both hubs.

Do they both use the same version of Pybricks? You can tell by looking in the bottom right of the app when the hub is connected.

**Reply by Devilly** [reply]
I just verified they are both running 3.2.3.

I will get back to you about the code. I'd like to give you a minimal reproducible example for it but just now I didn't manage to quickly get that up and running.

**Reply by Devilly** [reply]
### Latest

So, I'm going to leave below description here as reference but I've now see it working also on the Technic Hub. The interesting thing is this happened just because of running below script a couple times more. I have seen flakey behaviour concerning the bluetooth connection in the past while trying to connect from Pybricks Code (no clue what the cause was) so maybe this can be designated to the same cause?

### Previous findings

That fortunately didn't take too long. In both of below cases I expect to get the green button to shine but that only happens for the Inventor Hub. The Technic Hub hangs on the orange color.

Inventor Hub:

```python
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor, UltrasonicSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.geometry import Axis
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch

from usys import stdin, stdout

hub = InventorHub(Axis.X, Axis.Z)
hub.light.on(Color.ORANGE)

kp = float(stdin.buffer.read(1))

hub.light.on(Color.GREEN)

wait(2000)
```

Corresponding pc code:

```python
import asyncio
from time import time
from bleak import BleakScanner, BleakClient

UART_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
UART_RX_CHAR_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
UART_TX_CHAR_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"

HUB_NAME = "Robot Hub"

def hub_filter(device, ad):
    print(device.name)
    return device.name and device.name.lower() == HUB_NAME.lower()

def handle_disconnect(_):
    print("Hub was disconnected.")

def handle_rx(_, data: bytearray):
    pass

device = await BleakScanner.find_device_by_filter(hub_filter)
client = BleakClient(device, disconnected_callback=handle_disconnect)

async def send(client, data):
    await client.write_gatt_char(rx_char, data)

try:
    await client.connect()
    await client.start_notify(UART_TX_CHAR_UUID, handle_rx)
    nus = client.services.get_service(UART_SERVICE_UUID)
    rx_char = nus.get_characteristic(UART_RX_CHAR_UUID)

    await send(client, b"5.000")
    await send(client, b"0.010")
    await send(client, b"500.0")

    await asyncio.sleep(5)

except Exception as e:
    print(e)
finally:
    await client.disconnect()

``` 
-----

Technic Hub (same as above, just a different hub type):

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor, UltrasonicSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.geometry import Axis
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch

from usys import stdin, stdout

hub = TechnicHub(Axis.X, Axis.Z)
hub.light.on(Color.ORANGE)

kp = float(stdin.buffer.read(1))

hub.light.on(Color.GREEN)

wait(2000)
``` 

Corresponding pc code (same as above, just a different hub name):

```python
import asyncio
from time import time
from bleak import BleakScanner, BleakClient

UART_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
UART_RX_CHAR_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
UART_TX_CHAR_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"

HUB_NAME = "Technic Hub"

def hub_filter(device, ad):
    print(device.name)
    return device.name and device.name.lower() == HUB_NAME.lower()

def handle_disconnect(_):
    print("Hub was disconnected.")

def handle_rx(_, data: bytearray):
    pass

device = await BleakScanner.find_device_by_filter(hub_filter)
client = BleakClient(device, disconnected_callback=handle_disconnect)

async def send(client, data):
    await client.write_gatt_char(rx_char, data)

try:
    await client.connect()
    await client.start_notify(UART_TX_CHAR_UUID, handle_rx)
    nus = client.services.get_service(UART_SERVICE_UUID)
    rx_char = nus.get_characteristic(UART_RX_CHAR_UUID)

    await send(client, b"5.000")
    await send(client, b"0.010")
    await send(client, b"500.0")

    await asyncio.sleep(5)

except Exception as e:
    print(e)
finally:
    await client.disconnect()

```

**Reply by Devilly** [reply, marked as answer]
@laurensvalk it was a timing problem. Seems like the Technic Hub is somewhat slower in its execution and some randomly chosen waits I used were not long enough for the Technic Hub which caused it to shut down before being able to show the green light.
